### PR TITLE
[WIP] Caching, Attribution, Persistence and Theme selection features

### PR DESCRIPTION
**Pull request by marius-wieschollek**
This pull request adds some much requested features to the unsplash app. It is also extendable and can be used for future expansion. I also have a [docker setup](https://github.com/marius-wieschollek/unsplash/tree/docker) which you can use for testing. This pull request is mainly to ask @jancborchardt @ChristophWurst @auberginepop and @newhinton for comments

## #7 Fetch and cache images in advance
This feature will fetch a predefined amount of images and related information periodically (every three hours) via the unsplash api and store them locally. When the app is first installed or the db is empty, two images will be fetched immediately.

#### Pros
- More privacy friendly, since no CSP changes will be needed anymore and unsplash is not contacted by the user directly anymore
- More caching friendly since the image will be delivered once to the browser and then cached

#### Cons
- Less variety since the amount of cached images is limited

## User defined image theme
Users can choose the theme of the image to one of the predefined themes.
This could also be extended to cover the usecase of #4 .

## #9 Creator attribution and additional image information
grafik
On the login page and public pages, you will now find a small camera icon at the bottom left with the image description, creator information and link to the unsplash page for the image. If you're logged in this information will be in the notifications.

#### Pros
- Proper creator attribution
- Link to more details, likes and high-res version on unsplash.com

#### Cons
- Requires use of unsplash api with api key

## #24 Different image resolutions
Based on the width of your device, different image sizes are used via `@media` css queries.
Smartphones get small images, desktop machines get medium or large images.

## #2 Same image for the whole session
By default, the user will now have the same image for the whole session starting from the login page.
This can be changed in the settings. 

#### Pros
- Greatly reduces the amount of network traffic since the image is now cached

## Provider independent backend
The code is written in a way that allows the use of different image sources. This would allow the implementation of #5 and #6 .

## TODOs
- ~~#24 Allow high res option~~ done
- Some settings are supported in code, but not configurable in the ui
- Maybe allow configuration of the image caching time (currently 3 hours)

### CODE REVIEW

**Comment by ChristophWurst** [comment]
Lots of good stuff in this pull request, nice work :+1: 

I'm not a maintainer of this app, but usually it's better to separate multiple features into separate pull requests. That makes it *a lot* easier to review/discuss and the rejection of one feature doesn't block the other features as well.

File: lib/Services/AppSettingsService.php [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        $this->config->setAppValue($this->appName, self::USER_SUBJECT, $allowUserSubject);
+    }
+
+    /**
+     * Get the api key to use for the unsplash api
+     *
+     * @return string
+     */
+    public function getApiKey(): string {
+        return $this->config->getAppValue($this->appName, self::API_KEY, 'ed3b10c059c5da43fe16c34ce5467b6c472abc10d21c498a6159df9fb556930a');
```
Comment (newhinton) [inline lib/Services/AppSettingsService.php]: There is an api-key, it shouldn't be there, right? ;)

Comment (marius-wieschollek) [reply lib/Services/AppSettingsService.php]: Actually it is there on purpose. The api key is required, but if the user does not specify an api key a default one is used.
But i think that maybe i should hide it in the settings.

**Comment by marius-wieschollek** [comment]
I have now also actually read the [api guideline](https://medium.com/unsplash/unsplash-api-guidelines-28e0216e6daa), and there are some issues that need to be addressed before we can actually use it:

> You cannot use the Unsplash name directly in your application name and you cannot use the Unsplash logo as an app icon.

We do both.

> You cannot replicate the core user experience of Unsplash.
> [...] From the [details](https://medium.com/unsplash/unsplash-api-guidelines-replicating-unsplash-com-f505aa170f5b):
> 🚫 a wallpaper app returns Unsplash images for downloading. Without the integration, the app has no content.

That is essentially us. Maybe we can bypass that if we add more providers or argue that this is a Nextcloud Integration since this would be allowed.

> All API uses must use the hotlinked image URLs returned by the API under the photo.urls properties.

Which means no caching at all.

> When displaying a photo from Unsplash, your application must attribute Unsplash, the Unsplash photographer, and contain a link back to their Unsplash profile. All links back to Unsplash should use utm parameters in the ?utm_source=your_app_name&utm_medium=referral

We do add the attribution, but i need to add the parameters to the link.

> When your application performs something similar to a download (like when a user chooses the image to include in a blog post, set as a header, etc.), you must send a request to the download endpoint returned under the photo.links.download_location property.

If we want to give users the option to save the photo to their Nextcloud, we need to check that. 


If we want to do #7 we could use the regular random image url, but it is clearly not wanted by unsplash and i think we should respect that.

**Comment by marius-wieschollek** [comment]
@newhinton I have changed the api key behavior so that it will not update the api key if it is invalid. Instead it will show an error. Also if no custom api key is set the field will be empty.

**Comment by newhinton** [comment]
hm, i guess we're getting into trouble if we continue without major changes to the app.

@jancborchardt What do you say to the name/logo-problem?
I guess we need a new one?


We could just remove imagecaching, and replace it with link or id cashing, the browser would take care of cashing the images after the first time the image was loaded, and it would still enable the other features like single image for session and so on

**Comment by jancborchardt** [comment]
About the name and logo problem yep – that’s an issue. :)

Something like "Random Background" or "Random Image" is not as catchy, and the package and repository name would still need to be the same to keep the installed user base.

But yeah, I do not like how Unsplash went away from CC0 (Public Domain) to their own little license … if we use an API, I would much prefer Wikimedia Commons which will stay open.

**Comment by gsantner** [comment]
Is there some cache mechanism already on master? Not right?
Just remember that every F5 reload made a http request to unsplash with user IP..which I don't want.

Any chance this PR gets conflicts resolved and merged?

**Comment by marius-wieschollek** [comment]
unsplash does not allow downloading and caching images via their api

**Pull request closed without merging**